hack/tools/conversion-verifier/main.go

```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"go/types"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	storageVersion   = markers.Must(markers.MakeDefinition("kubebuilder:storageversion", markers.DescribesType, crdmarkers.StorageVersion{}))
+	nolintConversion = markers.Must(markers.MakeDefinition("kubebuilder:conversion:nolint", markers.DescribesType, false))
+
+	apiPackages         = map[string]map[string]*loader.Package{}
+	storageVersionTypes = map[string]*storageVersionType{}
+)
+
+type versionType struct {
+	base string
+	pkg  *loader.Package
+	typ  *markers.TypeInfo
+}
+
+func (v *versionType) IsConvertible() bool {
+	return hasMethod(v.pkg, v.typ, "ConvertFrom") && hasMethod(v.pkg, v.typ, "ConvertTo")
+}
+
+type storageVersionType struct {
+	*versionType
+
+	otherDecls map[string]*versionType
+}
+
+func (s *storageVersionType) IsHub() bool {
+	return hasMethod(s.pkg, s.typ, "Hub")
+}
+
+func main() {
+	var result error
+
+	// Define the marker collector.
+	col := &markers.Collector{
+		Registry: &markers.Registry{},
+	}
+	// Register the markers.
+	if err := col.Registry.Register(storageVersion); err != nil {
+		klog.Fatal(err)
+	}
+	if err := col.Registry.Register(nolintConversion); err != nil {
+		klog.Fatal(err)
+	}
+
+	// Load all packages.
+	packages, err := loader.LoadRoots("./...")
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	// First loop through all the packages
+	// and find the ones that have a storage version assigned.
+	for i := range packages {
+		root := packages[i]
```

Comment:
why is this called root? Is it because we're matching in l.91 to only act pm the /api packages and thus it's root for /api/v1alpha3, ..

Reply:
may be something in the line of
`apiFolderPath := parentFolder(apiPackage.PkgPath)` instead of `base := basePackage(root.PkgPath)`would help to read it?

Reply:
Refactored a bit, ptal